rules/cache/cache.go

```diff
@@ -0,0 +1,378 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cache
+
+import (
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	mid "github.com/m3db/m3collector/id"
+	"github.com/m3db/m3collector/rules"
+	mrules "github.com/m3db/m3metrics/rules"
+	"github.com/m3db/m3x/clock"
+)
+
+const (
+	numOngoingTasks          = 2
+	deletionThrottleInterval = 100 * time.Millisecond
+)
+
+var (
+	errCacheClosed = errors.New("cache is already closed")
+)
+
+type setType int
+
+const (
+	dontSetIfNotFound setType = iota
+	setIfNotFound
+)
+
+type sleepFn func(time.Duration)
+
+type elemMap map[mid.Hash]*element
+
+type results struct {
+	elems  elemMap
+	source rules.Source
+}
+
+func newResults(source rules.Source) results {
+	return results{elems: make(elemMap), source: source}
+}
+
+// cache is an LRU-based read-through cache.
+type cache struct {
+	sync.RWMutex
+
+	capacity          int
+	nowFn             clock.NowFn
+	sleepFn           sleepFn
+	freshDuration     time.Duration
+	stutterDuration   time.Duration
+	evictionBatchSize int
+	deletionBatchSize int
+	invalidationMode  InvalidationMode
+
+	namespaces map[string]results
+	list       lockedList
+	evictCh    chan struct{}
+	deleteCh   chan struct{}
+	toDelete   []elemMap
+	wgWorker   sync.WaitGroup
+	closed     bool
+	closedCh   chan struct{}
+}
+
+// NewCache creates a new cache.
+func NewCache(opts Options) rules.Cache {
```

Comment:
One thing I have a question about the entire Cache: is there a reason we give entries a TTL and then refresh them instead of simply evicting rules as the rules actually change due to explicit notifications from our KV watches?